src/Waitosaur.sol

```diff
@@ -0,0 +1,176 @@
+
+    event Locked(uint256 indexed amount);
+    event Unlocked();
+    event LockerUpdated(address indexed newLocker);
+    event UnLockerUpdated(address indexed newUnLocker);
+    event ReceiverUpdated(address indexed newReceiver);
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(
+        address owner_,
+        address _token,
+        address _locker,
+        address _unLocker,
+        address _receiver
+    ) public initializer {
+        if (_token == address(0)) revert InvalidTokenAddress();
+        __Ownable_init(owner_);
+        __UUPSUpgradeable_init();
+        StorageSlot.getUint256Slot(LOCKED_AMOUNT_STORAGE_SLOT).value = 0;
+        StorageSlot.getBooleanSlot(UNLOCKED_STORAGE_SLOT).value = true;
+        StorageSlot.getAddressSlot(ERC20_STORAGE_SLOT).value = _token;
+        StorageSlot.getAddressSlot(LOCKER_STORAGE_SLOT).value = _locker;
+        StorageSlot.getAddressSlot(UNLOCKER_STORAGE_SLOT).value = _unLocker;
+        StorageSlot.getAddressSlot(RECEIVER_STORAGE_SLOT).value = _receiver;
+    }
+
+    function lock(uint256 amount) external {
+        if (msg.sender != StorageSlot.getAddressSlot(LOCKER_STORAGE_SLOT).value)
+            revert NotLocker();
+        if (StorageSlot.getBooleanSlot(UNLOCKED_STORAGE_SLOT).value == false)
+            revert AlreadyLocked();
+        if (amount == 0) revert AmountZero();
+        StorageSlot.getUint256Slot(LOCKED_AMOUNT_STORAGE_SLOT).value = amount;
+        StorageSlot.getBooleanSlot(UNLOCKED_STORAGE_SLOT).value = false;
+        emit Locked(amount);
+        // Store the current block timestamp in LAST_LOCKED_STORAGE_SLOT
+        StorageSlot.getUint256Slot(LAST_LOCKED_STORAGE_SLOT).value = block
+            .timestamp;
+    }
+    function lastLocked() public view returns (uint256) {
+        return StorageSlot.getUint256Slot(LAST_LOCKED_STORAGE_SLOT).value;
+    }
+
+    function unlock() external {
+        if (
+            msg.sender !=
+            StorageSlot.getAddressSlot(UNLOCKER_STORAGE_SLOT).value
+        ) revert NotUnLocker();
+        if (StorageSlot.getBooleanSlot(UNLOCKED_STORAGE_SLOT).value == true)
+            revert AlreadyUnlocked();
+        if (StorageSlot.getUint256Slot(LOCKED_AMOUNT_STORAGE_SLOT).value == 0)
+            revert NothingLocked();
+        address tokenAddr = token();
+        (bool success, bytes memory data) = tokenAddr.call(
+            abi.encodeWithSignature("balanceOf(address)", address(this))
+        );
+        require(success && data.length >= 32, "balanceOf failed");
+        uint256 balance = abi.decode(data, (uint256));
+        if (
+            balance <
+            StorageSlot.getUint256Slot(LOCKED_AMOUNT_STORAGE_SLOT).value
+        ) revert InsufficientBalance();
+        if (
+            StorageSlot.getAddressSlot(RECEIVER_STORAGE_SLOT).value ==
+            address(0)
+        ) revert("Receiver not set");
```

Comment:
I guess it is better to check for this condition in the initializer and in appropriate update method